include/fuse_lowlevel.h

```diff
@@ -660,11 +662,11 @@ struct fuse_lowlevel_ops {
 	 * etc) in fi->fh, and use this in other all other directory
 	 * stream operations (readdir, releasedir, fsyncdir).
 	 *
-	 * Filesystem may also implement stateless directory I/O and not
-	 * store anything in fi->fh, though that makes it impossible to
-	 * implement standard conforming directory stream operations in
-	 * case the contents of the directory can change between opendir
-	 * and releasedir.
+	 * Implementing standard conforming directory stream operations when
+	 * the directory is modified between opendir and releasedir calls
+	 * requires either storing state in a handle indexed by fi->fh or using
+	 * readdir's off_t parameter as a cookie that will seek to the correct
+	 * entry in the directory even under concurrent modification.
```

Comment:
This sounds like you could implement standard conforming stream operations by using just `fi->fh`, but without using `off_t`. Is that what you mean? If so, I how would you do it? I don't see a way...

Reply:
That is what I meant! I'll clarify. :)

The simplest solution to POSIX-compliant readdir is to store the entire directory list in an array indexed by `fi->fh` and then using off_t as an integer offset into that array, but I just want to point out that it's possible to implement readdir correctly, even under concurrent modification, without storing state in `fi->fh` by ensuring that the off_t cookie always seeks to the correct place in the stream even under modification.

Reply:
I'm sorry, I appreciate your efforts, but I don't see why your version is better than the status quo (for example, you cannot generally "index" into a state). Could you remind me what the drawbacks of the current text are?

Reply:
Thanks for the feedback! I'm also not happy with the wording I chose. I'd welcome any suggestions.

The current text states that using `fi->fh` is required to implement POSIX-conforming readdir. This is not true - if off_t is treated as a magic 63-bit cookie that will always seek into the correct part of the directory stream, `fi->fh` is not needed. In fact, that's how filesystems like btrfs and xfs work. (And it's a good thing `fi->fh` is not required - should the FUSE_NO_OPENDIR_SUPPORT patch land, filesystems could opt out of the opportunity to even set it.)

Without being too verbose, I'd like to describe that it's possible to implement POSIX-conforming readdir by using each dirent's offset parameter as a cookie that indexes into a separate b-tree.

The easiest way to get POSIX conformance is to use `fi->fh` as a pointer to an array of N elements and have off_t merely be the index into that array. But it's certainly not required. It took me a while to figure this out myself, so I just wanted to save someone else the work of determining whether `fi->fh` is required for correctness.

Reply:
I see. How about removing this paragraph completely? The preceding paragraph already says "may", so it should be clear that this is not a must.

Instead, we could flesh out the description of readdir() and add something like: "If entries are added or removed during a readdir cycle, they may or may not be returned. However, they must not cause other entries to be skipped or returned more than once."

..and leave it to the user to figure out how the various ways to achieve that.

Reply:
(probably need a better phrase than "readdir cycle", something along the lines of "readdir getting an off_t value that it had returned previously but file were added/removed afterwards")